rfd/000?-database-access.md

```diff
@@ -0,0 +1,512 @@
+regular database clients they normally use to connect directly such as
+CLI clients (`psql`, `mysql`, etc.) as well as graphical interfaces (`pgAdmin`,
+`MySQL Workbench`, etc.).
+
+The use-case for this is to grant users access to a database in a transparent
+fashion, for example to let them do development in a test/stage environment
+or perform an emergency recovery on a production database instance using
+familiar tools.
+
+### Robot access
+
+The feature should be automation friendly so existing CI systems can take
+advantage of it.
+
+An example would be letting the tools like Ansible or Drone perform routine
+actions on a database such as migrations or backups and be able to audit it.
+
+### Programmatic access
+
+Programmatic access, as in configuring an application to talk to a database
+through Teleport proxy, should work automatically as long as it uses a driver
+that properly implements a particular database protocol and supports mutual
+TLS authentication.
+
+However, it is not the primary use-case, at least for the initial release,
+since it comes with a number of additional concerns and considerations such
+as performance requirements for high-traffic applications, automatic failover
+and so on.
+
+## Scope
+
+For the initial release we're focusing on supporting a single type of the
+database, PostgreSQL, with full protocol parsing.
+
+The following PostgreSQL deployment models are supported:
+
+* PostgreSQL instances deployed on-premises.
+* AWS RDS for PostgreSQL.
+* PostgreSQL-compatible AWS Aurora.
+
+The following features are provided:
+
+* Connecting to the database through the Teleport proxy, incl. trusted
+  clusters support.
+* Limiting access to database instances by labels with Teleport roles.
+* Limiting access to individual databases (within a particular database
+  instance) and database users.
+* Auditing of database connections and executed queries.
+
+## Architecture
+
+Architecturally Database Access is very similar to [Application Access](0008-application-access.md).
+Teleport gains a new mode of operation, "database service", similar to "ssh
+service", "application service" or a "kube service".
+
+Database service runs inside the customer private network and proxies database
+client connections received from the Teleport proxy to the target database. It
+is also responsible for protocol parsing of supported databases and authorizing
+requests coming from the proxy (i.e. granting/denying access to particular
+databases/database users).
+
+```
+              |‾‾‾‾‾‾‾|  reverse  |‾‾‾‾‾‾‾‾‾‾‾‾|          |‾‾‾‾‾‾‾‾‾‾|
+psql -------> | proxy | <-------- | db service | -------> | postgres |
+              |_______|   tunnel  |____________|          |__________|
+```
+
+The database client (such as psql) will talk to Teleport web proxy port (`3080`
+by default) which will use multiplexing to detect the database protocol and
+dispatch to an appropriate service.
```

Comment:
there is no problem from performance perspective because protocol switching uses first several bytes to differentiate protocols, this works fast

Reply:
But will we be able to guess the protocol reliably for other databases?

Reply:
I think since we want to consolidate the number of ports ideally to just 2 (k8s proxy + 443 for everything else) we don't have much choice but to try and do it, obviously "we don't know what we don't know" rule still stands so I guess we'll play it by ear.